### ISSUE TITLE
Expose method name in `Interceptor`s

### ISSUE THREAD

**Issue by Sushisource** [hyperium/tonic#772]
### Crates
tonic
### Motivation
I need to attach a deadline/timeout whose value varies based on which method I'm calling. Additionally, I implement some metrics like counting number of reqs/method, etc. Currently I can implement these as `tower` middlewares which totally works, but it'd be nice to be able to use the much simpler `Interceptor` interface for something this simple.

### Proposal
A `uri()` method available on the Request object like in an `http::Request` would be the simplest. Beyond that, it might be nice to do something more gRPC specific, like `.method()` return a `GrpcMethod` type which exposes `.service()` and `.method()` functions.

This is to avoid this hullabaloo:
`let maybe_method = req.uri().to_string().rsplitn(2, '/').next()`

### Alternatives
Right now I'm solving this in tower middleware. This can also be solved by manually building `Request`s before calling the generated client code, and inserting the appropriate metadata then.

The proposal would I think be a bit more convenient in many cases, though. (As a bonus, even easier if there was a way for me to get a static reference to the `GrpcMethod` (or `&'static str`) which defines the RPC path, currently just baked as a literal into the client code)

**Comment by LucioFranco** [comment]
I think this is the right place to use tower and I think we should make it easier to add load shedding to something like tonic.

**Comment by marcoferrer** [comment]
Ive noticed that in other gRPC runtimes, the interceptors get a reference to a sort of method descriptor that contains additional metadata such as rpc type. I think its worth including as well if this ever gets developed. Right now we're having a hard time adopting this library due to this issue + the reliance on prost and the lack of proto option support.

**Comment by LucioFranco** [comment]
> Ive noticed that in other gRPC runtimes, the interceptors get a reference to a sort of method descriptor that contains additional metadata such as rpc type. 

Yeah, I think we need to improve the metadata here from the prost side, the interceptors in general are not in a great spot right now. I think you can achieve all that you want with tower layers though?

> Right now we're having a hard time adopting this library due to this issue + the reliance on prost and the lack of proto option support.

Yes, I was just looking at that issue today for `prost`, its quite some serious design work and impl work required but if you all would like to offer some help I think we can get it implemented.

**Comment by CodingAnarchy** [comment]
This appears to be blocking something I'm trying to do with a basic authentication interceptor that I want on all services _except_ the `HealthCheck` service provided by `tonic-health`. Because I have no way to tell in the interceptor which service is being requested, I can't bypass the auth check for the `HealthCheck` service.

I don't really need the whole body and rewriting authentication as a full middleware is a lot more complex than the interceptor approach. It would be really nice if there was something available on the `Request` coming into the interceptor that gave us the `grpc.health.v1....` method path.

**Comment by George-Madeley** [comment]
Has this been resolved yet? Or has anyone found a work around? I have the same issue as @CodingAnarchy  bit I have authentication service that allows users to login, signup, delete account etc. Logging in should not require an access token but deleting an account would.
